File: lares/lararium_mcp/meme_graph.py
"""MemeGraph — adjacency structure over reachable memes in the carrier DAG.

Provides:
  Meme              — one resolved carrier node
  DeclaredUnresolved — a pranala edge whose target cannot be located on disk
  MemeGraph          — mutable adjacency structure; walk, sort, hash

All graph operations (topological_sort, detect_cycles, etc.) operate on the
memes already loaded into the graph.  I/O — reading carrier files and parsing
edges — belongs to the compiler; MemeGraph is a pure in-memory structure.
"""

from __future__ import annotations

import hashlib
import json
from collections import defaultdict, deque
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .pranala_parser import PranaEdge


# ---------------------------------------------------------------------------
# Meme
# ---------------------------------------------------------------------------

@dataclass
class Meme:
    uri: str
    path: Path | None
    content_hash: str
    metadata: dict[str, Any]
    edges_out: list[PranaEdge] = field(default_factory=list)
    virtual: bool = False
    exists: bool = False
    shape: Any = None  # CarrierShape when available; Any to avoid circular import

    @property
    def implements(self) -> tuple[str, ...]:
        """URIs of interfaces this meme declares via control/implements edges."""
        return tuple(
            e.to_uri for e in self.edges_out
            if e.role == 'implements' and e.family == 'control'
        )

    @classmethod
    def make_hash(cls, uri: str, file_bytes: bytes | None) -> str:
        payload = uri + ':' + (file_bytes.decode('utf-8', errors='replace') if file_bytes else 'virtual')
        return 'sha256:' + hashlib.sha256(payload.encode('utf-8')).hexdigest()


# ---------------------------------------------------------------------------
# DeclaredUnresolved
# ---------------------------------------------------------------------------

@dataclass
class DeclaredUnresolved:
    uri: str
    edge: PranaEdge
    severity: str  # "error" | "warning" | "info"

    @classmethod
    def from_edge(cls, edge: PranaEdge) -> 'DeclaredUnresolved':
        # Two-phase load pattern (Bazel/Flecs): interface attachment edges
        # (control/implements) point to abstract interface URIs that are
        # intentionally outside the boot walk spine.  They are not missing
        # boot dependencies — they are Phase 1 interface references that
        # resolve during interface_index construction.  Mark as "info".
        # Hard "error" reserves for structural edges (owns, composes) whose
        # targets must exist for the walk to be valid.
        if edge.family == 'control':
            if edge.role == 'implements':
                severity = 'info'
            else:
                severity = 'error'
        elif edge.family == 'relation':
            severity = 'warning'
        else:
            severity = 'info'
        return cls(uri=edge.to_uri, edge=edge, severity=severity)


# ---------------------------------------------------------------------------
# MemeGraph
# ---------------------------------------------------------------------------

class MemeGraph:
    """Mutable adjacency structure over carrier memes.

    Internal layout:
        memes:     uri → Meme
        adjacency: family → from_uri → [PranaEdge, ...]
    """

    def __init__(self) -> None:
        self.memes: dict[str, Meme] = {}
        # adjacency[family][from_uri] = [PranaEdge, ...]
        self.adjacency: dict[str, dict[str, list[PranaEdge]]] = defaultdict(lambda: defaultdict(list))

    # ------------------------------------------------------------------
    # Mutation
    # ------------------------------------------------------------------

    def add_meme(self, meme: Meme) -> None:
        """Insert a meme and index its outbound edges into the adjacency map."""
        self.memes[meme.uri] = meme
        for edge in meme.edges_out:
            self.adjacency[edge.family][edge.from_uri].append(edge)

    # ------------------------------------------------------------------
    # Query
    # ------------------------------------------------------------------

    def successors(self, uri: str, family: str) -> list[str]:
        """Outbound target URIs for a given meme and edge family."""
        return [e.to_uri for e in self.adjacency.get(family, {}).get(uri, [])]

    def edges_out(self, uri: str, family: str | None = None) -> list[PranaEdge]:
        """All outbound PranaEdge records from uri, optionally filtered by family."""
        meme = self.memes.get(uri)
        if meme is None:
            return []
        if family is None:
            return list(meme.edges_out)
        return [e for e in meme.edges_out if e.family == family]

    def one_hop_relation(self, control_uris: list[str] | set[str]) -> set[str]:
        """Relation-edge neighbours of control-reachable memes, excluding already-reachable."""
        seen = set(control_uris)
        additional: set[str] = set()
        for uri in control_uris:
            for target in self.successors(uri, 'relation'):
                if target not in seen:
                    additional.add(target)
                    seen.add(target)
        return additional

    # ------------------------------------------------------------------
    # Cycle detection
    # ------------------------------------------------------------------

    def detect_cycles(self, families: list[str] | None = None) -> list[list[str]]:
        """DFS gray/black cycle detection over given families (default: all loaded)."""
        if families is None:
            families = list(self.adjacency.keys())

        WHITE, GRAY, BLACK = 0, 1, 2
        color: dict[str, int] = defaultdict(int)
        violations: list[list[str]] = []
        stack_path: list[str] = []

        def visit(uri: str, family: str) -> None:
            if color[uri] == GRAY:
                cycle_start = stack_path.index(uri)
                violations.append(stack_path[cycle_start:] + [uri])
                return
            if color[uri] == BLACK:
                return
            color[uri] = GRAY
            stack_path.append(uri)
            for target in self.successors(uri, family):
                visit(target, family)
            stack_path.pop()
            color[uri] = BLACK

        for family in families:
            # Skip mutual-pair check for relation family (permitted by pranala law)
            color.clear()
            for uri in list(self.adjacency.get(family, {}).keys()):
                if color[uri] == WHITE:
                    visit(uri, family)

        return violations

    # ------------------------------------------------------------------
    # Topological sort (Kahn's algorithm)
    # ------------------------------------------------------------------

    def topological_sort(
        self,
        uri_set: set[str] | list[str],
        family: str = 'control',
    ) -> list[str]:
        """Kahn's BFS topological sort restricted to uri_set over family edges.

        Returns URIs in sources-before-targets order.
        """
        nodes = set(uri_set)
        in_degree: dict[str, int] = {u: 0 for u in nodes}
        successors_map: dict[str, list[str]] = {u: [] for u in nodes}

        for uri in nodes:
            for target in self.successors(uri, family):
                if target in nodes:
                    in_degree[target] += 1
                    successors_map[uri].append(target)

        queue: deque[str] = deque(u for u in nodes if in_degree[u] == 0)
        result: list[str] = []
        while queue:
            uri = queue.popleft()
            result.append(uri)
            for target in successors_map[uri]:
                in_degree[target] -= 1
                if in_degree[target] == 0:
                    queue.append(target)

        # Any remaining nodes (cycle residue) appended in stable order
        remaining = [u for u in nodes if u not in set(result)]
        result.extend(sorted(remaining))
        return result

    # ------------------------------------------------------------------
    # Closure hash
    # ------------------------------------------------------------------

    def closure_hash(self, uri_list: list[str], edge_list: list[PranaEdge]) -> str:
        """SHA256 of sorted meme hashes + serialised edge records."""
        meme_hashes = sorted(
            self.memes[u].content_hash for u in uri_list if u in self.memes
        )
        edge_records = sorted(
            (e.from_socket, e.to_uri, e.family, e.role or '')
            for e in edge_list
        )
        payload = json.dumps({'memes': meme_hashes, 'edges': edge_records},
                             sort_keys=True, separators=(',', ':'))
        return 'sha256:' + hashlib.sha256(payload.encode('utf-8')).hexdigest()

    # ------------------------------------------------------------------
    # Declared-unresolved scan
    # ------------------------------------------------------------------

    def declared_unresolved(self) -> list[DeclaredUnresolved]:
        """Edges pointing to URIs not loaded in this graph.

        Deduplicates by target URI, keeping the highest severity seen.
        Severity order: error > warning > info.
        """
        _rank = {'error': 2, 'warning': 1, 'info': 0}
        best: dict[str, DeclaredUnresolved] = {}
        for meme in self.memes.values():
            for edge in meme.edges_out:
                if edge.to_uri not in self.memes:
                    du = DeclaredUnresolved.from_edge(edge)
                    existing = best.get(du.uri)
                    if existing is None or _rank[du.severity] > _rank[existing.severity]:
                        best[du.uri] = du
        return sorted(best.values(), key=lambda d: (-_rank[d.severity], d.uri))

File: packages/lararium-core/src/meme-graph.ts
/**
 * meme-graph — pure adjacency structure over reachable memes in the lares DAG.
 *
 * Isomorphic: no fs/path/DOM imports. I/O belongs to lararium-node.
 * No CarrierShape dependency — shape lives in the projection layer.
 * MemeRating derived at ingest boundary and stored on MemeRecord, not on Meme.
 *
 * Schema: lar:///ha.ka.ba/@lares/api/v0.1/lararium/schema/meme-graph
 */

import type { PranalaEdge } from "./pranala-parser.js";
import { utf8Bytes, sha256Hex, defaultCryptoProvider } from "./crypto.js";

// ---------------------------------------------------------------------------
// MemeRating — derived at ingest, stored on MemeRecord
// ---------------------------------------------------------------------------

/**
 * Rating assigned at the ingest boundary when a meme is put into the store.
 * Never stored on the graph node — the graph operates on `Meme` (structural),
 * not on ingest-boundary judgments.
 *
 * "kapu"  — restricted; access control applies; do not traverse without auth check
 * "ano"   — anomalous; failed one or more validation rules at ingest
 * "meme"  — standard carrier; passes all checks; default for valid corpus memes
 * "data"  — pure data tiddler; no control/reaction edges expected or validated
 * "noise" — unrecognized structure; skipped by graph traversal
 */
export type MemeRating = "kapu" | "ano" | "meme" | "data" | "noise";

// ---------------------------------------------------------------------------
// MemeRecord — ingest boundary output (replaces CarrierRecord)
// ---------------------------------------------------------------------------

/**
 * Output of the ingest boundary: a meme validated and rated, ready for the graph.
 * Immutable by convention — the graph receives it as a snapshot.
 */
export interface MemeRecord {
  readonly uri:        string;
  readonly metadata:   Record<string, unknown>;
  readonly implements: readonly string[];
  readonly edges:      readonly PranalaEdge[];
  readonly rating:     MemeRating;
}

// ---------------------------------------------------------------------------
// Meme — graph node (pure structure, no I/O, no shape)
// ---------------------------------------------------------------------------

export interface Meme {
  uri:          string;
  /** Relative path within lares/memes/ — null for virtual or hostful memes. */
  laresRelPath: string | null;
  contentHash:  string;
  metadata:     Record<string, unknown>;
  edgesOut:     PranalaEdge[];
  virtual:      boolean;
  exists:       boolean;
}

// ---------------------------------------------------------------------------
// DeclaredUnresolved — edge that references a URI not present in the graph
// ---------------------------------------------------------------------------

export interface DeclaredUnresolved {
  uri:      string;
  edge:     PranalaEdge;
  severity: "error" | "warning" | "info";
}

export function declaredUnresolvedFromEdge(edge: PranalaEdge): DeclaredUnresolved {
  let severity: "error" | "warning" | "info";
  if (edge.family === "control") {
    severity = edge.role === "implements" ? "info" : "error";
  } else if (edge.family === "relation") {
    severity = "warning";
  } else {
    severity = "info";
  }
  return { uri: edge.toUri, edge, severity };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Returns the interface URIs this meme composes (control:implements edges). */
export function memeImplements(meme: Meme): string[] {
  return meme.edgesOut
    .filter((e) => e.role === "implements" && e.family === "control")
    .map((e) => e.toUri);
}

/** Returns the single parent class URI (control:extends edge), or undefined if absent. */
export function memeExtends(meme: Meme): string | undefined {
  return meme.edgesOut
    .find((e) => e.role === "extends" && e.family === "control")
    ?.toUri;
}

/**
 * Compute a stable content hash for a meme.
 * Uses the URI + raw bytes as the preimage so the hash identifies both location and content.
 * null bytes → virtual meme (no disk content).
 */
export async function makeMemeHash(uri: string, fileBytes: Uint8Array | null): Promise<string> {
  const content = fileBytes ? new TextDecoder().decode(fileBytes) : "virtual";
  const payload = uri + ":" + content;
  return "sha256:" + await sha256Hex(utf8Bytes(payload), defaultCryptoProvider);
}

// ---------------------------------------------------------------------------
// MemeGraph — pure adjacency structure
// ---------------------------------------------------------------------------

export class MemeGraph {
  readonly memes = new Map<string, Meme>();

  // adjacency.get(family)?.get(fromUri) → PranalaEdge[]
  private readonly adjacency = new Map<string, Map<string, PranalaEdge[]>>();

  // Reverse index: interfaceUri → Set<memeUri> (for memesByInterface — control:implements)
  private readonly _byInterface = new Map<string, Set<string>>();
  // Reverse index: parentUri → Set<memeUri> (for memesByParent — control:extends)
  private readonly _byParent = new Map<string, Set<string>>();

  addMeme(meme: Meme): void {
    this.memes.set(meme.uri, meme);
    for (const edge of meme.edgesOut) {
      // Forward adjacency
      let familyMap = this.adjacency.get(edge.family);
      if (!familyMap) { familyMap = new Map(); this.adjacency.set(edge.family, familyMap); }
      let list = familyMap.get(edge.fromUri);
      if (!list) { list = []; familyMap.set(edge.fromUri, list); }
      list.push(edge);

      // Reverse implements index (control:implements → interface conformance)
      if (edge.family === "control" && edge.role === "implements") {
        let set = this._byInterface.get(edge.toUri);
        if (!set) { set = new Set(); this._byInterface.set(edge.toUri, set); }
        set.add(meme.uri);
      }
      // Reverse extends index (control:extends → single class parent)
      if (edge.family === "control" && edge.role === "extends") {
        let set = this._byParent.get(edge.toUri);
        if (!set) { set = new Set(); this._byParent.set(edge.toUri, set); }
        set.add(meme.uri);
      }
    }
  }

  // ---------------------------------------------------------------------------
  // Forward traversal
  // ---------------------------------------------------------------------------

  successors(uri: string, family: string): string[] {
    return (this.adjacency.get(family)?.get(uri) ?? []).map((e) => e.toUri);
  }

  edgesOut(uri: string, family?: string): PranalaEdge[] {
    const meme = this.memes.get(uri);
    if (!meme) return [];
    return family ? meme.edgesOut.filter((e) => e.family === family) : [...meme.edgesOut];
  }

  // ---------------------------------------------------------------------------
  // Graph queries
  // ---------------------------------------------------------------------------

  /**
   * Expand a set of control URIs by one hop in the `relation` family.
   * Returns the input set PLUS all relation-adjacent URIs (one hop only).
   */
  oneHopRelation(controlUris: Iterable<string>): Set<string> {
    const seen = new Set(controlUris);
    const additional = new Set<string>();
    for (const uri of seen) {
      for (const target of this.successors(uri, "relation")) {
        if (!seen.has(target)) { additional.add(target); seen.add(target); }
      }
    }
    return additional;
  }

  /**
   * All memes whose `control:implements` edge targets the given interface URI.
   * Returns an empty set if no meme implements the interface.
   */
  memesByInterface(interfaceUri: string): ReadonlySet<string> {
    return this._byInterface.get(interfaceUri) ?? new Set();
  }

  /**
   * All memes whose `control:extends` edge targets the given parent URI.
   * Returns direct children only — use `allTransitiveDeps` for full inheritance chains.
   */
  memesByParent(parentUri: string): ReadonlySet<string> {
    return this._byParent.get(parentUri) ?? new Set();
  }

  /**
   * Full transitive closure of successors from `rootUri` over `family` edges.
   * Returns all URIs reachable (excluding the root itself unless it is in a cycle).
   */
  allTransitiveDeps(rootUri: string, family = "control"): Set<string> {
    const visited = new Set<string>();
    const stack = [rootUri];
    while (stack.length > 0) {
      const uri = stack.pop()!;
      for (const next of this.successors(uri, family)) {
        if (!visited.has(next)) {
          visited.add(next);
          stack.push(next);
        }
      }
    }
    return visited;
  }

  /**
   * Resolved closure: all URIs transitively reachable from any of `rootUris`
   * via `control` edges, INCLUDING the roots, filtered to memes that exist.
   */
  resolvedClosure(rootUris: Iterable<string>): Set<string> {
    const result = new Set<string>();
    for (const root of rootUris) {
      result.add(root);
      for (const dep of this.allTransitiveDeps(root, "control")) {
        result.add(dep);
      }
    }
    return result;
  }

  /**
   * Topological sort of a subset of nodes (`nodeSet`) over edges in `family`.
   * Uses Kahn's algorithm. Returns nodes in dependency order (roots first).
   * Any cycles in the family will cause affected nodes to be omitted from the result;
   * call `detectCycles()` separately to surface them.
   */
  topologicalSort(nodeSet: ReadonlySet<string>, family: string): string[] {
    // Build in-degree map within the subset
    const inDegree = new Map<string, number>();
    for (const uri of nodeSet) inDegree.set(uri, 0);

    for (const uri of nodeSet) {
      for (const next of this.successors(uri, family)) {
        if (nodeSet.has(next)) {
          inDegree.set(next, (inDegree.get(next) ?? 0) + 1);
        }
      }
    }

    const queue: string[] = [];
    for (const [uri, deg] of inDegree) {
      if (deg === 0) queue.push(uri);
    }

    const result: string[] = [];
    while (queue.length > 0) {
      const uri = queue.shift()!;
      result.push(uri);
      for (const next of this.successors(uri, family)) {
        if (!nodeSet.has(next)) continue;
        const newDeg = (inDegree.get(next) ?? 1) - 1;
        inDegree.set(next, newDeg);
        if (newDeg === 0) queue.push(next);
      }
    }

    return result;
  }

  /**
   * All edges in the graph that point to URIs not present in this graph.
   * The severity follows `declaredUnresolvedFromEdge` rules:
   *   control (non-implements) → "error", relation → "warning", others → "info".
   */
  declaredUnresolved(): DeclaredUnresolved[] {
    const result: DeclaredUnresolved[] = [];
    for (const meme of this.memes.values()) {
      for (const edge of meme.edgesOut) {
        if (!this.memes.has(edge.toUri)) {
          result.push(declaredUnresolvedFromEdge(edge));
        }
      }
    }
    return result;
  }

  // ---------------------------------------------------------------------------
  // Cycle detection
  // ---------------------------------------------------------------------------

  /**
   * DFS cycle detection over the given families (default: all families in the graph).
   * Returns an array of cycle paths — each path is the list of URIs forming the cycle,
   * with the first URI repeated at the end: [A, B, C, A].
   */
  detectCycles(families?: string[]): string[][] {
    const fams = families ?? [...this.adjacency.keys()];
    const violations: string[][] = [];

    for (const family of fams) {
      const WHITE = 0, GRAY = 1, BLACK = 2;
      const color = new Map<string, number>();
      const stackPath: string[] = [];

      const visit = (uri: string): void => {
        if (color.get(uri) === GRAY) {
          const cycleStart = stackPath.indexOf(uri);
          violations.push([...stackPath.slice(cycleStart), uri]);
          return;
        }
        if (color.get(uri) === BLACK) return;
        color.set(uri, GRAY);
        stackPath.push(uri);
        for (const next of this.successors(uri, family)) {
          visit(next);
        }
        stackPath.pop();
        color.set(uri, BLACK);
      };

      for (const uri of this.memes.keys()) {
        if (!color.has(uri)) visit(uri);
      }
    }

    return violations;
  }
}

